### PR TITLE
DS-3489: Read-only search rest endpoint

### PR DESCRIPTION
**Pull request by tomdesair**
This PR implements the contract of https://github.com/DSpace/Rest7Contract/blob/master/search-endpoint.md related to ticket https://jira.duraspace.org/browse/DS-3489.

We still plan to add some unit tests, but the base code is ready for reviewing.

Since is the first complex Rest Controller next to `org.dspace.app.rest.RestResourceController`, we refactored the code a bit so that all HAL links are created by a`org.dspace.app.rest.link.HalLinkFactory`. For each HAL resource, there is a corresponding Link factory that will add the required links to that resource. This has the following advantages:
* **Separation of concerns**: Only HAL link factories are responsible for adding HAL links
* **Open for extension, closed for modification**: Adding extra links to a HAL resource is as simple as defining a new factory (e.g. when writing DSpace customisations or when implementing authorization). Spring will autowire the factory and apply it to all applicable resources without the need of modifying any additional code.
* **Enable projections to filter links**: This should also make it possible for a projection to filter-out certain HAL links in the future.

Example test calls can be found in the examples of the related contract (https://github.com/DSpace/Rest7Contract/pull/13).

### CODE REVIEW

File: dspace-api/pom.xml [hunk 1]
```diff
@@ -354,6 +354,10 @@
             <groupId>commons-collections</groupId>
             <artifactId>commons-collections</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-collections4</artifactId>
```
Comment (tdonohue) [inline dspace-api/pom.xml]: While I don't object to it, I'd like to ensure both of these versions of `commons-collections` are necessary. I notice we already have it just above, and this PR adds in `commons-collections4`.

File: dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java [hunk 2]
```diff
@@ -538,7 +540,7 @@ private synchronized static Flyway setupFlyway(DataSource datasource)
      * @throws SQLException if database error
      *      If database cannot be upgraded.
      */
-    public static synchronized void updateDatabase()
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java]: Note: There are *two* `updateDatabase()` methods (with different params) in this class, and it seems you only removed the `synchronized` from one of them. Is that on purpose?

Comment (tomdesair) [reply dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java]: I think that currently we synchronise our threads a bit too much. I think this results in all REST API calls to go sequentially through this lock. Additionally, we might think about not checking Flyway for each and every REST API call. But I'll take that discussion to another ticket. I've reverted the change here.

File: dspace-api/pom.xml (hunk 1 above)
Comment (tomdesair) [reply dspace-api/pom.xml]: Collections4 offers some new calls like `emptyIfNull`. I agree that we should upgrade all our `commons-collections` usages to version 4, but I think we should do that in a separate ticket.

File: dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java (hunk 2 above)
Comment (tomdesair) [reply dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java]: I've created a ticket for this: https://jira.duraspace.org/browse/DS-3781

File: dspace-api/pom.xml (hunk 1 above)
Comment (tomdesair) [reply dspace-api/pom.xml]: I've created this ticket to address this: https://jira.duraspace.org/browse/DS-3782

**Review by lap82 — APPROVED** [review]
Ok, thank you very much for this PR. It's look good. I tested it and it works.

**Review by terrywbrady — COMMENTED** [review]
The code is clean and well-organized.  My review has been very cursory at this point.  It would take some time to look over all of the functionality in a single review.  I am inclined to recommend a merge.  If there is a specific area to focus my review, please let me know.

**Pull request merged**